keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,556 @@
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Logging Interface](#logging-interface)
+    - [Logger initialization](#logger-initialization)
+    - [Storing Logger and metadata in Context](#storing-logger-and-metadata-in-context)
+    - [Example](#example)
+  - [New logging formats](#new-logging-formats)
+    - [Json format](#json-format)
+  - [Logging configuration](#logging-configuration)
+  - [Kubernetes log schema](#kubernetes-log-schema)
+  - [Migration](#migration)
+    - [Stage 1 / Alpha](#stage-1--alpha)
+    - [Preparation](#preparation)
+    - [Automatic migration](#automatic-migration)
+    - [Stage 2 / Beta](#stage-2--beta)
+    - [Stage 3 / GA](#stage-3--ga)
+    - [Stage 4 / Deprecation](#stage-4--deprecation)
+  - [User Stories](#user-stories)
+    - [Story 1 - Kubernetes developer working with local logs](#story-1---kubernetes-developer-working-with-local-logs)
+    - [Story 2 - Cluster Administrator configures log ingestion into ElasticSearch](#story-2---cluster-administrator-configures-log-ingestion-into-elasticsearch)
+    - [Story 3 - Administrator uses Elasticsearch to debug problem with workload](#story-3---administrator-uses-elasticsearch-to-debug-problem-with-workload)
+  - [Performance](#performance)
+  - [Risks and Mitigations](#risks-and-mitigations)
+    - [Migration failing halfway](#migration-failing-halfway)
+    - [Hard to customize logger format](#hard-to-customize-logger-format)
+    - [Huge increase of log volume](#huge-increase-of-log-volume)
+- [Design Details](#design-details)
+  - [Mapping between logging methods](#mapping-between-logging-methods)
+  - [Example of migrating klog call](#example-of-migrating-klog-call)
+  - [Graduation Criteria](#graduation-criteria)
+    - [Alpha](#alpha)
+    - [Beta](#beta)
+    - [GA](#ga)
+    - [Removal of old logging format](#removal-of-old-logging-format)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+    - [Enhance klog instead of replacing it with logr](#enhance-klog-instead-of-replacing-it-with-logr)
+    - [Avoid invoking both klog and logr api](#avoid-invoking-both-klog-and-logr-api)
+- [Code organisation / infrastructure needed](#code-organisation--infrastructure-needed)
+- [Production Readiness Review Process](#production-readiness-review-process)
+<!-- /toc -->
+
+## Summary
+
+Current logging in the Kubernetes control-plane doesn’t provide any uniform structure for logs which makes storing, processing, querying and analyzing logs hard and forces administrators and developers to rely on ad-hoc solutions. This KEP proposes a new logging API and a mechanism for standardizing logging keys that makes logs much easier to ingest and process by modern logging solutions.
+
+## Motivation
```

Comment:
IMO, nothing in this motivation section justifies the enormity of the proposed change. What is the compelling reason to do this?

Reply:
I think the key points are:

> but it’s not enough for cluster operators to do any analysis on large multi-master cluster, find recurring problems on fleet or implement reliable monitoring and alerting on top of it.

and

> logs in order to be useful needs also to be searchable meaning that it should be easy for the user to find the most relevant information in them

But they're a bit buried amongst the background and exposition.  Perhaps we could highlight them, or call them out in a TL;DR, or something?

From my perspective, the core of the motivation is more or less this:

It's *way* harder than it has any right to be to search through k8s logs at any sort of (even relatively small) scale.  Things like the API server request logs require quite a bit of massaging with regular expressions, fuzzy matching, or even per-resource filtering before you can reliably search through them.  Following containers through Kubelet logs (in my experience) requires quite a bit of "filter this, see where you're at, filter that, keep this, realize that one line uses a different format or accidentally matches your regex, try again".  Similar for many controller logs, especially when you mix controller and client-go logs.  This makes it harder to follow the logs, identify issues, locate anomalies, etc.

This proposal significantly reduces this burden -- massaging log lines into searchable data is eliminated in favor of pre-structuring the log for consumption.  Regexes are reduced or eliminated in favor of filters on specific fields.  Inconsistent nomenclature is discouraged through helpers, linting, and guidelines, making following a specific object or series of objects a *lot* easier.

Reply:
> the API server request logs require quite a bit of massaging

Yes, but you should use the audit logs, which don't have these problems.

Also fixing this would be a 1-10 line change.

> Following containers through Kubelet logs

I don't have recent personal experience doing that, but if the Kubelet logs use inconsistent ways of identifying a container, maybe that should just be fixed? I don't think changing the whole logging interface is on the most direct route to solving that problem?

Reply:
> Yes, but you should use the audit logs, 

(which are nice because they're structured)

> which don't have these problems.

Sure, that's definitely viable in some cases.  In others, you want log lines from other parts of the apiserver too, though.  Manually interleaving audit logs with component logs as I'm debugging doesn't sound fun to me.

> Also fixing this would be a 1-10 line change. 

I'm not sure how this is the case.  Pulling meaning out of non-structured logs *always* requires some sort of relatively complex logic, even if the logs are consistent (which is, in-and-of-itself, hard to enforce with format strings, especially between components).  Sentence-style messages effectively mean you don't have a field separator, so identifying distinct pieces of data needs regexes or similar, and attaching semantic meaning to line requires a-priori knowing what the line is saying .  Plus, filtering on that particular line becomes difficult, because you have to be able to generally say what it looks like, as opposed to asking for matches to a constant message.

> but if the Kubelet logs use inconsistent ways of identifying a container, maybe that should just be fixed
 
a) it still doesn't solve the problem of how to simplify pulling information out of log lines.  You could say "kubelet always uses `[$CONTAINER] $MSG`, but then you've just got poorly done per component hard-to-ingest semi-structured logs (and you can't easily pull out other pieces of data).

b) this isn't just about one part of API server, or just the pod lifecycle code in kubelet -- from what I've experienced, many different parts of the system have different conventions, inconsistent formats, and varied nomenclature between components and within a given component.  Those two were just examples.

Reply:
My criticism isn't that the proposal isn't an improvement. It's that it's not worth SWE-years to make this change when we can make things tolerable with SWE days. Making the apiserver log line easily greppable is a 1 hour task. Fixing kubelet logs to be consistent is probably a lot harder than that, but still doable.

IOW, I'm saying, don't solve the general problem when you can solve the major specific problems at 1/100th the effort.

Reply:
E.g. this took 10 minutes: https://github.com/kubernetes/kubernetes/pull/87203 (and probably needs another 10 minutes of test fixing, but I have to run)